Bluetooth: host: Change directed advertising to privacy-enabled peer

joerchan opened the pull request:
Change the advertising option that controls if the directed advertiser
will use an RPA or the identity address of the peer for the initiator
address.
This option currently has two issues:
 - It behaves differently if the privacy feature has been enabled,
   which can be confusing for application to use.
 - It cannot start a directed advertiser towards a peer that is not
   privacy-enabled and has distributed an IRK.

This commit includes the following changes:
 - When privacy has been enabled in order to advertise towards a
   privacy-enabled peer the BT_LE_ADV_OPT_DIR_ADDR_RPA option must now
   be set (same as when privacy has been disabled).
 - It is now possible to start a directed advertiser using the identity
   address of the peer when privacy is enabled.
 - When privacy has been enabled the advertising option combination
   of using the local identity address and an RPA as the initiator
   address is now disallowed and will return an error code.
   This is done because this combination did not actually work and would
   have used the identity address of the peer instead.
 - If the controller does not support controller-based privacy then
   using the option BT_LE_ADV_OPT_DIR_ADDR_RPA will return ENOTSUP
   because this behavior cannot be done with host-based privacy.

joerchan committed:
Bluetooth: host: Change directed advertising to privacy-enabled peer

Change the advertising option that controls if the directed advertiser
will use an RPA or the identity address of the peer for the initiator
address.
This option currently has two issues:
 - It behaves differently if the privacy feature has been enabled,
   which can be confusing for application to use.
 - It cannot start a directed advertiser towards a peer that is not
   privacy-enabled and has distributed an IRK.

This commit includes the following changes:
 - When privacy has been enabled in order to advertise towards a
   privacy-enabled peer the BT_LE_ADV_OPT_DIR_ADDR_RPA option must now
   be set (same as when privacy has been disabled).
 - It is now possible to start a directed advertiser using the identity
   address of the peer when privacy-enabled.
 - When privacy has been enabled the advertising option combination
   of using the local identity address and an RPA as the initiator
   address is now disallowed and will return an error code.
   This is done because this combination did not actually work and would
   have used the identity address of the peer instead.
 - If the controller does not support controller-based privacy then
   using the option BT_LE_ADV_OPT_DIR_ADDR_RPA will return ENOTSUP
   because this behavior cannot be done with host-based privacy.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: shell: Fix shell scan option duplicate filtering options

Fix the shell always setting the filter duplicates option for active
scan.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

MarekPieta reviewed (approved):
I verified that it works fine with nrf_desktop devices in various usecases. Thank you for the fix.

joerchan committed:
doc: release-notes: Add note about directed advertising address option

Added note about change in the option for advertising directed
towards a privacy enabled peer.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -7818,16 +7831,8 @@ static int le_adv_set_random_addr(struct bt_le_ext_adv *adv, uint32_t options,
-				/* This will not use RPA for our own address
-				 * since we have set zeroed out the local IRK.
-				 */
+			if (dir_adv && options & BT_LE_ADV_OPT_DIR_ADDR_RPA) {
```
Minor: please put `()` around bit-wise operations.

joerchan replied:
Even when it is the only thing in the if?

jhedberg replied:
Not then, since then you'd have `(( ... ))`. Here however you're combining it with other stuff.

joerchan replied:
Yes I am :) fixed.
